jvm-packages/xgboost4j-spark/src/main/scala/org/apache/spark/SparkParallelismTracker.scala

```
@@ -0,0 +1,122 @@
+import org.codehaus.jackson.map.ObjectMapper
+
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future, TimeoutException}
+
+/**
+ * A tracker that ensures enough number of executor cores are alive.
+ * Throws an exception when the number of alive cores is less than nWorkers.
+ *
+ * @param sc The SparkContext object
+ * @param timeout The maximum time to wait for enough number of workers.
+ * @param nWorkers nWorkers used in an XGBoost Job
+ */
+class SparkParallelismTracker(
+    val sc: SparkContext,
+    timeout: Long,
+    nWorkers: Int) {
+
+  private[this] val mapper = new ObjectMapper()
+  private[this] val logger = LogFactory.getLog("XGBoostSpark")
+  private[this] val url = sc.uiWebUrl match {
+    case Some(baseUrl) => new URL(s"$baseUrl/api/v1/applications/${sc.applicationId}/executors")
+    case _ => null
+  }
+  private[spark] var failedReason: String = null
+
+  private[this] def numAliveCores: Int = {
+    try {
+      mapper.readTree(url).findValues("totalCores").asScala.map(_.asInt).sum
+    } catch {
+      case ex: Throwable =>
+        logger.warn(s"Unable to read total number of alive cores from REST API." +
+          s"Health Check will be ignored.")
+        ex.printStackTrace()
+        Int.MaxValue
+    }
+  }
+
+  private[this] def waitForCondition(
+      condition: => Boolean,
+      timeout: Long,
+      checkInterval: Long = 100L) = {
+    val monitor = Future {
+      while (!condition) {
+        Thread.sleep(checkInterval)
+      }
+    }
+    Await.ready(monitor, timeout.millis)
+  }
+
+  private[this] def safeExecute[T](body: => T): T = {
+    sc.listenerBus.listeners.add(0, new ExecutorLostListener(this))
+    try {
+      val result = body
+      if (failedReason != null) {
+        throw new SparkException(s"ExecutorLost during XGBoost Training: $failedReason")
```

what if body does not return due to a single task failure (by default 4 failure will lead to a failed job), I think you should throw exception in listener which will cause SparkContext shutdown automatically

---

Spark will swallow NonFatal exception in listener, so you have to throw interruptedexception (as an example)